paddle/operators/math/depthwise_conv.cu

```diff
@@ -0,0 +1,337 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/operators/math/depthwise_conv.h"
+#include "paddle/platform/cuda_helper.h"
+
+namespace paddle {
+namespace operators {
+namespace math {
+
+// A Cuda kernel to compute the depthwise convolution forward pass
+// in NCHW format.
+template <typename T>
+__global__ void KernelDepthwiseConv(
+    const int nthreads, const T* const input_data, const T* const filter_data,
+    const int batch_size, const int output_channels, const int output_height,
+    const int output_width, const int input_channels, const int input_height,
+    const int input_width, const int filter_multiplier, const int filter_height,
+    const int filter_width, const int stride_height, const int stride_width,
+    const int padding_height, const int padding_width, T* const output_data) {
+  int index = (blockIdx.x * gridDim.y + blockIdx.y) * blockDim.x + threadIdx.x;
+
+  if (index < nthreads) {
+    const int batch = index / output_channels / output_height / output_width;
+    const int c_out = (index / output_height / output_width) % output_channels;
+    const int h_out = (index / output_width) % output_height;
+    const int w_out = index % output_width;
+
+    const int c_in = c_out / filter_multiplier;
+    const T* weight = filter_data + c_out * filter_height * filter_width;
+    T value = 0;
+    const int h_in_start = -padding_height + h_out * stride_height;
+    const int w_in_start = -padding_width + w_out * stride_width;
+    const int h_in_end =
+        -padding_height + h_out * stride_height + filter_height - 1;
+    const int w_in_end =
+        -padding_width + w_out * stride_width + filter_width - 1;
+
+    const int in_offset =
+        ((batch * input_channels + c_in) * input_height) * input_width;
+
+    if ((h_in_start >= 0) && (h_in_end < input_height) && (w_in_start >= 0) &&
+        (w_in_end < input_width)) {
+      for (int kh = 0; kh < filter_height; ++kh) {
+        for (int kw = 0; kw < filter_width; ++kw) {
+          const int h_in = h_in_start + kh;
+          const int w_in = w_in_start + kw;
+          const int offset = in_offset + h_in * input_width + w_in;
+
+          value += (*weight) * input_data[offset];
+          ++weight;
+        }
+      }
+    } else {
+      for (int kh = 0; kh < filter_height; ++kh) {
+        for (int kw = 0; kw < filter_width; ++kw) {
+          const int h_in = h_in_start + kh;
+          const int w_in = w_in_start + kw;
+          if ((h_in >= 0) && (h_in < input_height) && (w_in >= 0) &&
+              (w_in < input_width)) {
+            const int offset = in_offset + h_in * input_width + w_in;
+            value += (*weight) * input_data[offset];
+          }
+          ++weight;
+        }
+      }
```

Comment:
It seems that line53~77 can be written more shortly and more efficiently. 
Improving the efficiency by reducing the number of judgments.

e.g.
```
      h_in_end = ...
      w_in_end = ...
      for (int kh = h_in_start; kh < h_in_end; ++kh) {
        for (int kw = w_in_start; kw < w_in_end; ++kw) {
           ...
           ...
       }
     }
```

Reply:
If so, we have to add extra complex judgment for filter data. This may outweigh the benefits.

Reply:
I think the following code can work, and it is not complex.
```
      h_end = h_in_start + filter_height > input_height? input_height:h_in_start + filter_height;
      w_end = w_in_start + filter_width > input_width? input_width:w_in_start + filter_width;
      h_start = h_in_start > 0? h_in_start:0;
      w_start = w_in_start > 0? w_in_start:0;
      input_data += in_offset;
      for (int h_in = h_start; h_in < h_end; ++h_in) {
        for (int w_in = w_start; w_in < w_end; ++w_in) {
            const int offset = h_in * input_width + w_in;
            value += weight[(h_in - h_start)*filter_width + w_in] * input_data[offset];
          }
        }
      }
```